docs/dotnet/dotnet-kestrel-metrics.md

```diff
@@ -0,0 +1,452 @@
+<!--- Hugo front matter used to generate the website version of this page:
+linkTitle: Kestrel
+--->
+
+# Semantic Conventions for Kestrel web server metrics
+
+**Status**: [Stable][DocumentStatus]
+
+This article defines semantic conventions for Kestrel web server.
+
+<!-- toc -->
+
+- [Kestrel endpoint](#kestrel-endpoint)
+- [Metric: `kestrel.active_connections`](#metric-kestrelactive_connections)
+- [Metric: `kestrel.connection.duration`](#metric-kestrelconnectionduration)
+- [Metric: `kestrel.rejected_connections`](#metric-kestrelrejected_connections)
```

Comment:
One theme I'm noticing here (and I realize this is what .NET currently produces, so not suggestion a change):

There's a lot of these UpDownCounters with `active_*`, `queued_*`, etc.

In OTEL semantic conventions, we tend to try to use labels with meaningful sums where appropriate, e.g. the `usage` and `time` conventions:


However, the overall rule is:

> ["As a rule of thumb, aggregations over all the attributes of a given metric SHOULD be meaningful,"](https://prometheus.io/docs/practices/naming/#metric-names) as Prometheus recommends.


My question for .NET (as a naive, non .NET user) ->  Should `active`, `rejected`, `queued` and `upgraded` be labels on a `connections` metric, or is this not meaningful when aggregated?

Reply:
Great point. 

I see the following opportunity: `kestrel.active_connections`, `kestrel.queued_connections` and maybe ``kestrel.upgraded_connections` could be grouped together (they are updowncounters with the same unit and set of attributes). Rejected connection is a total counter and can't be reported along with them.

These 3 could be measured with something like `kestrel.current.connections` or `kestrel.connections` and additional attribute like `kestrel.connection.state` = `active` | `queued`.

If my understanding of upgrades is correct, then upgraded connections are first counted in active ones and upgrade happens after the connection is established. If we added it to `kestrel.connections`, it'd need to be expressed by a different attribute with some questionable logic:
- `-1` updowncounter when upgrade is done
- `+1` immediately with an extra attribute like `http.protocol.initial_version = 1.1`  and new value of `network.protocol.version = 2`

It looks a bit unusual and surprising, so I'm not sure if `kestrel.upgraded_connections` makes sense in this group. 

Having said that, this ship has already sailed, but I created this issue https://github.com/dotnet/aspnetcore/issues/52437 for .NET team for the reference and to consider changing it in the future.

Reply:
/fyi @JamesNK